src/coreclr/src/vm/interoplibinterface.cpp

```diff
@@ -630,94 +634,121 @@ namespace
+                    identity,
+                    GetCurrentCtxCookie(),
+                    gc.objRefMaybe->GetSyncBlockIndex(),
+                    flags);
+
+                if (uniqueInstance)
                 {
-                    // Failed to set the context; one must already exist.
-                    // Remove from the cache above as well.
+                    extObjCxt = resultHolder.GetContext();
+                }
+                else
+                {
+                    // Attempt to insert the new context into the cache.
                     ExtObjCxtCache::LockHolder lock(cache);
-                    cache->Remove(resultHolder.GetContext());
+                    extObjCxt = cache->FindOrAdd(identity, resultHolder.GetContext());
+                }
+
+                // If the returned context matches the new context it means the
+                // new context was inserted or a unique instance was requested.
+                if (extObjCxt == resultHolder.GetContext())
+                {
+                    // Update the object's SyncBlock with a handle to the context for runtime cleanup.
+                    SyncBlock* syncBlock = gc.objRefMaybe->GetSyncBlock();
+                    InteropSyncBlockInfo* interopInfo = syncBlock->GetInteropInfo();
+                    _ASSERTE(syncBlock->IsPrecious());
+
+                    // Since the caller has the option of providing a wrapper, it is
+                    // possible the supplied wrapper already has an associated external
+                    // object and an object can only be associated with one external object.
+                    if (!interopInfo->TrySetExternalComObjectContext((void**)extObjCxt))
+                    {
+                        // Failed to set the context; one must already exist.
+                        // Remove from the cache above as well.
+                        ExtObjCxtCache::LockHolder lock(cache);
+                        cache->Remove(resultHolder.GetContext());
+
+                        COMPlusThrow(kNotSupportedException);
+                    }
 
-                    COMPlusThrow(kNotSupportedException);
+                    // Detach from the holder to avoid cleanup.
+                    (void)resultHolder.DetachContext();
+                    STRESS_LOG2(LF_INTEROP, LL_INFO100, "Created EOC (Unique Instance: %d): 0x%p\n", (int)uniqueInstance, extObjCxt);
                 }
 
-                // Detach from the holder to avoid cleanup.
-                (void)resultHolder.DetachContext();
-                STRESS_LOG2(LF_INTEROP, LL_INFO100, "Created EOC (Unique Instance: %d): 0x%p\n", (int)uniqueInstance, extObjCxt);
+                _ASSERTE(extObjCxt->IsActive());
             }
-
-            _ASSERTE(extObjCxt->IsActive());
         }
 
         GCPROTECT_END();
 
-        RETURN gc.objRef;
+        *objRef = gc.objRefMaybe;
+        RETURN (gc.objRefMaybe != NULL);
     }
+
+    bool g_IsGlobalComWrappersRegistered;
```

Comment:
I believe there are other global variables in this file. Can we keep them together?